src/pydvl/influence/frameworks/torch_differentiable.py

```diff
@@ -2,36 +2,231 @@
 Contains all parts of pyTorch based machine learning model.
 """
 import logging
-from typing import Any, Callable, List, Sequence, Tuple
+from typing import Any, Callable, Dict, List, Optional, Sequence, Tuple
 
 import numpy as np
 import torch
 import torch.nn as nn
 from torch import autograd
 from torch.autograd import Variable
+from torch.utils.data import DataLoader
 
 from ...utils import maybe_progress
 from .twice_differentiable import TwiceDifferentiable
 
 __all__ = [
     "TorchTwiceDifferentiable",
+    "solve_linear",
+    "solve_batch_cg",
+    "solve_lissa",
+    "as_tensor",
+    "stack",
+    "cat",
+    "einsum",
+    "mvp",
 ]
 logger = logging.getLogger(__name__)
 
 
-def flatten_gradient(grad):
+def flatten_all(grad) -> torch.Tensor:
     """
     Simple function to flatten a pyTorch gradient for use in subsequent calculation
     """
     return torch.cat([el.reshape(-1) for el in grad])
 
 
-def solve_linear(matrix: torch.Tensor, b: torch.Tensor):
-    """Computes the solution of a square system of linear equations"""
-    return torch.linalg.solve(matrix, b)
+def solve_linear(
+    model: TwiceDifferentiable,
+    training_data: DataLoader,
+    b: torch.Tensor,
+    *,
+    lam: float = 0.0,
```

Comment:
`lam` seems a bit unintuitive for a parameter name. Why not `hessian_perturbation` or `regularizer`. I'd prefer the former because it's explicit about the type of regularization. Note that this should be added to the documentation, since it's $(H + \lambda I) x = b$ that is solved.